### PR TITLE
Generate and encode stable edge IDs when importing OSM directly

### PR DESCRIPTION
**Pull request by mark-idleman**
For moving completely to a GH-based OSM import (ie, phasing out the R5->GH pipeline), this will mimic our currently stable edge ID encoding scheme found [here](https://github.com/replicahq/model-r5/blob/swl_master/src/main/java/com/conveyal/r5/streets/EdgeStore.java#L213-L231). The goal is to properly create and encode stable edge IDs for new OSM imports in GH, while not re-computing the IDs when loading a graph from disk (which presumably should already have these IDs encoded).

**Note:** This is not meant to be merged until we're certain we'll be using a GH-only import setup; once we get to that point, this is one of the main changes that will need to be merged.

### CODE REVIEW

**Comment by mark-idleman** [comment]
@michaz a few questions for you on this one: 

1. Is this the best (ie, "least hacky possible") way of doing this? 
2. If so, is there any easy way to check/set a flag of some sort to tell if we're importing OSM from scratch (in which case we want the stable IDs to be calculated) vs loading an existing graph (in which case we don't want to re-compute IDs)? This check happens inside `importOrLoad` but I couldn't think of a good way of mimicking it while also calling the function.
3. Is there any way to get the OSM way type info from an existing GH edge? The current stable ID implementation uses this (with [these](https://github.com/replicahq/model/blob/master/mapservice/api/network_link_references.py#L23-L35) types) but searches in the GH forums seem to indicate it's hard/not possible

File: web-bundle/src/main/java/com/graphhopper/swl/StableIdEncodedValues.java [hunk 1]
```diff
@@ -43,17 +46,39 @@ final String getStableId(boolean reverse, EdgeIteratorState edge) {
         return Long.toUnsignedString(Longs.fromByteArray(stableId));
     }
 
-    final void setStableId(boolean reverse, EdgeIteratorState edge, String stableId) {
```
Comment (mark-idleman) [inline web-bundle/src/main/java/com/graphhopper/swl/StableIdEncodedValues.java:46]: note: we weren't calling this at all before, so changing the parameter list shouldn't be an issue.

File: web-bundle/src/main/java/com/graphhopper/swl/StableIdEncodedValues.java [hunk 2]
```diff
@@ -43,17 +46,39 @@ final String getStableId(boolean reverse, EdgeIteratorState edge) {
...
 
         UnsignedIntEncodedValue[] idBytes = reverse ? reverseStableIdEnc : stableIdEnc;
         for (int i=0; i<8; i++) {
-            edge.set(idBytes[i], Byte.toUnsignedInt(stableIdBytes[i]));
+            edge.set(idBytes[i], Byte.toUnsignedInt(stableId[i]));
         }
     }
 
+    private static byte[] calculateStableEdgeId(EdgeIteratorState edge) {
+        // todo: how to get type of street?
```
Comment (kaelgreco) [inline web-bundle/src/main/java/com/graphhopper/swl/StableIdEncodedValues.java:61]: this is the OSM type, right? I can't fully remember what we use. Honestly, we could use our own type if that's easier. I couldn't find a "standard" implementation

File: web-bundle/src/main/java/com/graphhopper/swl/StableIdEncodedValues.java [hunk 3]
```diff
@@ -43,17 +46,39 @@ final String getStableId(boolean reverse, EdgeIteratorState edge) {
...
+        String hashString = "Reference";
+
+        PointList geometry = edge.fetchWayGeometry(FetchMode.ALL);
+        GHPoint3D baseNode = geometry.get(edge.getBaseNode());
+        GHPoint3D adjNode = geometry.get(edge.getAdjNode());
+        double startLat = baseNode.getLat();
+        double startLong = baseNode.getLon();
+        double endLat = adjNode.getLat();
+        double endLong = adjNode.getLon();
+        hashString += String.format(" %.5f %.5f %.5f %.5f", startLong, startLat, endLong, endLat);
```
Comment (kaelgreco) [inline web-bundle/src/main/java/com/graphhopper/swl/StableIdEncodedValues.java:71]: Do you remember what precision we write csv at? Are the hashes consistent?

**Review by kaelgreco — COMMENTED** [review]
cool! just so i understand: we won't be using this now, but idea now is to mimic stable edge id encoding in GH and then switch to loading osm in GH so we're ready to 🔪  r5?

File: web-bundle/src/main/java/com/graphhopper/swl/StableIdEncodedValues.java (hunk 3 above)
Comment (mark-idleman) [reply web-bundle/src/main/java/com/graphhopper/swl/StableIdEncodedValues.java:71]: It looks like in the street edges CSVs we write lat/longs with 6 places after the decimal...and [we hash using the same number as well](https://github.com/replicahq/model/blob/master/mapservice/api/network_link_references.py#L58-L73)

However, the [original shared streets implementation](https://github.com/sharedstreets/sharedstreets-builder/blob/master/src/main/java/io/sharedstreets/data/SharedStreetsReference.java#L331-L346) only uses 5 places, which is why I chose this I think. Should I switch it to 6 to keep it consistent with our current hashing scheme?

File: web-bundle/src/main/java/com/graphhopper/swl/StableIdEncodedValues.java (hunk 2 above)
Comment (mark-idleman) [reply web-bundle/src/main/java/com/graphhopper/swl/StableIdEncodedValues.java:61]: Yeah, that's what we use right now - it's actually just an int corresponding to the OSM way type I think: https://github.com/replicahq/model/blob/master/mapservice/api/network_link_references.py#L66

I guess, if this information isn't easily accessible here, we just need to make sure that what we choose to hash with is sufficient to guarantee all of edges will have unique hashes...it seems like this specific part of the hash string deals with the case where links of different types have complete overlap in their geometries, so we'll need to accommodate that somehow here using the info available in GH edges

**Comment by mark-idleman** [comment]
> cool! just so i understand: we won't be using this now, but idea now is to mimic stable edge id encoding in GH and then switch to loading osm in GH so we're ready to 🔪 r5?

Exactly right 👍 Once we're ready to 🔪 r5 (and go directly from OSM -> GH), we'll need to have stable IDs encoded in GH instead. In the near-term, we're hoping to build the nationwide routers right way using this OSM->GH method, so solidifying the hash scheme now will keep that consistent when we switch over the regional routers in the future too.

File: web-bundle/src/main/java/com/graphhopper/swl/StableIdEncodedValues.java (hunk 2 above)
Comment (mark-idleman) [reply web-bundle/src/main/java/com/graphhopper/swl/StableIdEncodedValues.java:61]: OK, I actually messed around some more and figured out how to grab this info from a GH edge (luckily, it's encoded by default). The options are [here](https://github.com/graphhopper/graphhopper/blob/dfdd92cf5e0d7f2c6c2289b98b646dfb6cfd36c1/core/src/main/java/com/graphhopper/routing/ev/RoadClass.java#L25), they seem to line up pretty well with the current enum we're using for road type.

**Comment by mark-idleman** [comment]
~~@kaelgreco I just realized I've been overlooking the [actual Java implementation we currently use in R5](https://github.com/replicahq/model-r5/blob/swl_master/src/main/java/com/conveyal/r5/streets/EdgeStore.java#L217-L231), and just going off of the python implementation I posted above 🤦 Not a big deal, but this solidifies some details I'll copy over when writing the new hash.~~

~~I have no context on why we re-hash the MD5 with farmhash - should I copy that setup here?~~

7/10 all of the above were addressed, and dealt with in 8fb1fae.

**Review by kaelgreco — APPROVED** [review]

**Pull request merged**